client.go

```diff
@@ -255,34 +255,39 @@ func (c *client) State() ClientState {
 func (c *client) setState(state ClientState) {
 	c.mx.Lock()
 	defer c.mx.Unlock()
+
 	c.state = state
 	_ = c.dbg.Log("state", state)
+
 	for _, ch := range c.stateChangeChans {
-		go func(ch chan struct{}) {
+		go func(ch chan ClientState, state ClientState) {
 			c.mx.Lock()
 			defer c.mx.Unlock()
+
 			for _, cch := range c.stateChangeChans {
```

Comment:
@philippseith why is this double loop needed? I noticed that the tests error with send on closed channel but why is that? Is there a hidden race condition somewhere?

Reply:
Because ch meanwhile might have been removed from stateChangeChans by cancelObserveStateChanged.